homeassistant/components/permobil/__init__.py

```diff
@@ -0,0 +1,66 @@
+"""The MyPermobil integration."""
+from __future__ import annotations
+
+import logging
+
+from mypermobil import MyPermobil, MyPermobilClientException
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    CONF_CODE,
+    CONF_EMAIL,
+    CONF_REGION,
+    CONF_TOKEN,
+    CONF_TTL,
+    Platform,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+
+from .const import APPLICATION, COORDINATOR, DOMAIN
+from .coordinator import MyPermobilCoordinator
+
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up MyPermobil from a config entry."""
+    default: dict = {
+        COORDINATOR: {},
+    }
+    hass.data.setdefault(DOMAIN, default)
+
+    # create the API object from the config and save it in hass
+    session = hass.helpers.aiohttp_client.async_get_clientsession()
+    p_api = MyPermobil(
+        application=APPLICATION,
+        session=session,
+        email=entry.data.get(CONF_EMAIL),
+        region=entry.data.get(CONF_REGION),
+        code=entry.data.get(CONF_CODE),
+        token=entry.data.get(CONF_TOKEN),
+        expiration_date=entry.data.get(CONF_TTL),
+    )
+    try:
+        p_api.self_authenticate()
+    except MyPermobilClientException as err:
```

Comment:
Does it throw an specific exception for authentication error? As in, can we know the difference between an connection issue and an authentication issue?

If so, we can raise an auth error that will trigger a re-auth flow for the end-user.

Reply:
This will only raise an exception if the any of the config data is wrong (Basically this can only be caused by the expiration_date being old). The checks are done locally by the client and does not make and requests to the API. 

The actual requests to the API are done in the coordinator which to my understanding will handle connection errors automatically. If the API itself returns an error (such as the token being invalid for whatever reason) it currently raises `UpdateFailed` perhaps that should be something different. 
See line 56 in coordinator.py

Reply:
Seems to me we should raise `ConfigEntryAuthFailed` instead so we initate the reauth flow to fix whatever is the problem?

Reply:
Do you mean to except whatever exception that `async_config_entry_first_refresh` could raise and then raise a `ConfigEntryAuthFailed` from that exception?

Reply:
```
try:
    await coordinator.async_config_entry_first_refresh()
except e:
    raise ConfigEntryAuthFailed from e
```

like so? @gjohansson-ST

Reply:
No, I mean that if `p_api.self_authenticate()` raises and error we can assume (?) it's an auth issue and should therefore raise `ConfigEntryAuthFailed` and implement a reauth flow.

Further down you have `coordinator.async_config_entry_first_refresh()` which will raise `ConfigEntryNotReady` if there is problem fetching data and will then retry to load the config entry a bit later.